### PR TITLE
Persist sidebar show state across views

### LINKED ISSUE #152: Sidebar state
It is annoying that the sidebar resets and pops in/out whenever a user navigates to a new page

- [x] Remember whether the sidebar what collapsed or not. This needs to take into account the current logic that shows/hides the sidebar based on the reference/library and screen size.
- [ ] If navigating from a reference item to another reference item via the sidebar, it would be great if the sidebar doesn't reload and goes to top of list (not sure whether this is possible with Gatsby routing)

### PR DESCRIPTION
**Pull request by bravomartin**
Modified the sidebar hook to persist the show state in `sessionStorage`. 

As for item 2 in the checklist, the only way to persist the scroll position is to do it in javascript. Found [this post](https://levelup.gitconnected.com/persist-and-remember-page-scroll-position-i-e-window-scrolly-using-react-hooks-f80884211f2d) that could be adapted to do this, but I'm not sure it's necessary. It already feels A LOT better with this change. @runemadsen thoughts? 

Fixes #152

### CODE REVIEW

**Review by runemadsen — CHANGES_REQUESTED** [review]
Awesome! A few changes are needed:

- If you have not interacted with the sidebar, you still get this odd opening animation on every page navigation. Try to delete your `sessionStorage` items, go to a reference item, and click the sidebar to go to another reference item. You will see the sidebar jumping out. When you have hidden/shown the sidebar, it no longer does this.
- The code doesn't work on `npm run build`. It's throwing a `"window" is not available during server side rendering.`. Please also check that the sidebar logic works on the built site, since the `sessionStorage` logic often works on dev but not in the the production build.

**Comment by bravomartin** [comment]
I made _some_ progress but I still get errors. I remember we had similar issues with euromomo. It has to do with hydrating or some optimization gatsby is doing.

If you see these screenshots, you can see that the data is actually being updated but the rendering  is only partially triggered. 

I'll continue to debug tomorrow.

(notice strokeWeight=2 and the handlers showing the modified path)

image

Screen Shot 2021-06-30 at 11 48 35 PM

**Comment by runemadsen** [comment]
Hmm, weird. What is the problem with the frontpage sketch? I thought I fixed all of that yesterday?

**Comment by bravomartin** [comment]
After trying every possible combination I could imagine, I landed at this: We attempt to figure out the open state on the first render, which causes the server to get confused for a second on first  open, and a separate `animate` state is set in both Sidebar and ContentWithSidebar at `300ms`. This causes a re-render that fixes the mismatch.


Hiding elements on null approach causes a pretty significant flash on every route change.  
Basically any attempt to have the server and browser match would either cause weird rendering issues, some odd flashing, or the animation to trigger on every route change.

@runemadsen give it a try and let me know what you think.

**Comment by runemadsen** [comment]
Thanks! This doesn't seem to be working for me. The sidebar state is not saved across views, and I'm still getting the animating effect on load. I'll try to take a look at it today.

**Comment by runemadsen** [comment]
Okay, I just rewrote it a bit and it seems to do the trick.

I simplified the `useSidebar` hook so it shows the sidebar as default. Since we're using a simple boolean as default, we will have the same markup rendered on the server and the client, which is important in order to not have a mismatch in the first place. On mount, the component will then check whether the sidebar should be hidden (if the screen is too small or `sessionStorage` is set to `false`). I also overloaded the `setShowSidebar` function to sync to `sessionStorage` since the `useEffect` was triggered on mount and caused weird issues.

@bravomartin I think this does the trick, but take a look and let me know if you find any issues.

**Comment by bravomartin** [comment]
ok done. take a look and merge

**Comment by runemadsen** [comment]
Screen Shot 2021-07-06 at 8 50 47 AM

@bravomartin This code doesn't run for me. I can fix it tho.

**Comment by runemadsen** [comment]
Screen Shot 2021-07-06 at 9 04 04 AM

It seems like the `ContentWithSidebar` functionality is now broken. I'm trying to look into it.

**Comment by runemadsen** [comment]
Okay, got it to work. I think we're good to go!
